### DISCUSSION TITLE
EntityManager.schema wrongly replacing default schema at runtime

### DISCUSSION THREAD

**Discussion by nicomouss** [mikro-orm/mikro-orm#5167] [General]
### Describe the bug

At runtime, the wildcard schema is replaced with `EntityManager.schema`, but it also wrongly replace the default schema.
Please see reproduction code below.

### Reproduction

Let's say I have a `User` entity defined in the default schema, and a `Trip` entity defined with the wilcard schema

```ts
// User entity
class User {
    id: number;
    name: string;
    constructor(name: string) {
        this.name = name;
    }
}

// Trip entity
class Trip {
    id: number;
    title: string;
    constructor(title: string) {
        this.title = title;
    }
}

// Note: I always use `EntitySchema` to configure my mappings, to keep my entities as clean as possible.
// However, the pb described here is the same if you use decorators.

// EntitySchema for User entity
const userEntitySchema = new EntitySchema<User>({
    class: User,
    properties: {
        id: { primary: true, type: Number },
        name: { type: String },
    }
});

// EntitySchema for Trip entity
const tripEntitySchema = new EntitySchema<Trip>({
    class: Trip,
    properties: {
        id: { primary: true, type: Number },
        title: { type: String },
    },
    schema: "*"
});
```

Now when I do this:

```ts
// Initializing the orm
const orm = await MikroORM.init({
    entities: [userEntitySchema, tripEntitySchema],
    driver: PostgreSqlDriver,
    host: ...,
    port: ...,
    dbName: ...,
    user: ...,
    password: ...,
    schemaGenerator: {
        disableForeignKeys: false,
    }
});

await orm.schema.createSchema();
await orm.schema.createSchema({ schema: "schema1" });

await orm.em.transactional(async (em) => {
    // Targeting specific schema
    em.schema = "schema1"; 

    // Creating new entities
    const trip = new Trip("Patagonia");
    const user = new User("Lisa");

    // Persisting/Flushing the entities - it breaks for User entity !
    await em.persistAndFlush([trip, user]);
});
```

This breaks when mikro-orm flushes the `User` entity, with the following error:

```bash
TableNotFoundException: insert into "schema1"."user" ("name") values ('Lisa') returning "id" - relation "schema1.user" does not exist
```

Proper generated SQL for the `User` entity should be:

```sql
insert into "user" ("name") values ('Lisa') returning "id"
```

Note that the generated SQL is ok for the `Trip` entity, as it properly targets the right schema.
The pb only occurs if some entities are defined in the default schema, like the `User` entity above.

Also note that if I explicitly define the default schema for the `User` entity like below, then everything is fine:

```ts
// EntitySchema for User entity
const userEntitySchema = new EntitySchema<User>({
    class: User,
    properties: {
        id: { primary: true, type: Number },
        name: { type: String },
    },
    schema: "default"
});
```

Generated SQL in that case is correct:

```sql
insert into "public"."user" ("name") values ('Lisa') returning "id"
```

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.0.5

### Node.js version

18

### Operating system

MacOs

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
I believe this is actually the correct behavior and changing this would be breaking, `em.schema` is the [default schema name](https://mikro-orm.io/docs/multiple-schemas#default-schema-on-entitymanager), it's an alternative for using the `schema` option in the ORM config (as well as using `FindOptions.schema` on call site), those two should be working the same. If you don't specify the `schema` in your entity definition, it means the entity belongs to the default schema, which can be overridden by the ORM config as well as `em.schema` and `FindOptions.schema`, those options are not "only for wildcard schemas", they are much older than that feature (although `em.schema` is a bit newer). If you want to fix the entity to one schema, put it explicitly to the entity definition.

**Reply by nicomouss** [reply]
Sure, that makes sense. Thanks for clarification.
